File: src/deepseek/torch/utils/device.py
"""
Unified Device Selection for DeepSeek PyTorch

This module provides centralized device selection with CUDA → MPS → CPU priority chain.

Features:
- Automatic device detection with consistent priority (CUDA first)
- Configurable device selection via environment variables
- Binary search auto batch size optimization
- Device information and availability checking

Usage:
    from deepseek.torch.utils.device import get_device, DeviceConfig, AutoBatchSizer
    
    # Get best available device (CUDA → MPS → CPU)
    device = get_device()
    
    # With custom configuration
    config = DeviceConfig.from_env()
    device = get_device_with_config(config)
    
    # Auto batch sizing
    sizer = AutoBatchSizer(AutoBatchConfig(max_memory_mb=8000))
    optimal_batch = sizer.find_optimal_batch_size_binary(memory_per_sample_mb=100)
"""

import os
from dataclasses import dataclass
from enum import Enum

import torch

from deepseek.torch.utils.logging import get_logger

logger = get_logger(__name__)


# =============================================================================
# Device Priority Enum
# =============================================================================

class DevicePriority(Enum):
    """Device selection priority."""
    CUDA_FIRST = "cuda_first"  # CUDA → MPS → CPU (recommended for training)
    MPS_FIRST = "mps_first"    # MPS → CUDA → CPU (for Apple Silicon optimization)
    CPU_ONLY = "cpu_only"      # Always use CPU


# =============================================================================
# Device Configuration
# =============================================================================

@dataclass
class DeviceConfig:
    """Configuration for device selection.
    
    Attributes:
        priority: Device selection priority order
        cuda_device_id: CUDA device ID for multi-GPU systems
        mps_device_id: MPS device ID (typically 0)
        min_batch_size: Minimum batch size for auto-adjustment
        max_batch_size: Maximum batch size for auto-adjustment
    """
    priority: DevicePriority = DevicePriority.CUDA_FIRST
    cuda_device_id: int = 0
    mps_device_id: int = 0
    min_batch_size: int = 1
    max_batch_size: int = 256
    
    @classmethod
    def from_env(cls) -> "DeviceConfig":
        """Create configuration from environment variables.
        
        Environment variables:
            DEEPSEEK_DEVICE_PRIORITY: "cuda", "mps", or "cpu"
            DEEPSEEK_CUDA_DEVICE: CUDA device ID (default: 0)
            DEEPSEEK_MIN_BATCH_SIZE: Minimum batch size (default: 1)
            DEEPSEEK_MAX_BATCH_SIZE: Maximum batch size (default: 256)
        """
        priority_str = os.environ.get("DEEPSEEK_DEVICE_PRIORITY", "cuda").lower()
        if priority_str == "mps":
            priority = DevicePriority.MPS_FIRST
        elif priority_str == "cpu":
            priority = DevicePriority.CPU_ONLY
        else:
            priority = DevicePriority.CUDA_FIRST
        
        try:
            cuda_device_id = int(os.environ.get("DEEPSEEK_CUDA_DEVICE", "0"))
        except ValueError:
            cuda_device_id = 0
        
        try:
            min_batch_size = int(os.environ.get("DEEPSEEK_MIN_BATCH_SIZE", "1"))
        except ValueError:
            min_batch_size = 1
        
        try:
            max_batch_size = int(os.environ.get("DEEPSEEK_MAX_BATCH_SIZE", "256"))
        except ValueError:
            max_batch_size = 256
        
        return cls(
            priority=priority,
            cuda_device_id=cuda_device_id,
            min_batch_size=min_batch_size,
            max_batch_size=max_batch_size,
        )
    
    def get_priority_order(self) -> list[str]:
        """Get the priority order as a list of device type names."""
        if self.priority == DevicePriority.CUDA_FIRST:
            return ["cuda", "mps", "cpu"]
        elif self.priority == DevicePriority.MPS_FIRST:
            return ["mps", "cuda", "cpu"]
        else:
            return ["cpu"]


# =============================================================================
# Auto Batch Size Configuration
# =============================================================================

@dataclass
class AutoBatchConfig:
    """Configuration for automatic batch size adjustment.
    
    Attributes:
        max_memory_mb: Maximum memory budget in MB
        memory_fraction: Fraction of max memory to use (0.0-1.0)
        min_batch_size: Minimum batch size
        max_batch_size: Maximum batch size
        auto_adjust: Whether to enable auto-adjustment
    """
    max_memory_mb: float = 8000  # 8 GB default
    memory_fraction: float = 0.9
    min_batch_size: int = 1
    max_batch_size: int = 256
    auto_adjust: bool = True


class AutoBatchSizer:
    """Automatic batch size optimizer using binary search.
    
    Uses binary search to find the optimal batch size that fits within
    the memory budget. Faster convergence than linear adjustment.
    
    Example:
        config = AutoBatchConfig(max_memory_mb=8000)
        sizer = AutoBatchSizer(config)
        optimal = sizer.find_optimal_batch_size_binary(memory_per_sample_mb=100)
    """
    
    def __init__(self, config: AutoBatchConfig | None = None):
        self.config = config or AutoBatchConfig()
    
    def find_optimal_batch_size_binary(self, memory_per_sample_mb: float) -> int:
        """Find optimal batch size using binary search.
        
        Args:
            memory_per_sample_mb: Memory required per sample in MB
            
        Returns:
            The optimal batch size within configured bounds
        """
        if not self.config.auto_adjust:
            return self.config.max_batch_size
        
        budget_mb = self.config.max_memory_mb * self.config.memory_fraction
        
        # Binary search for optimal batch size
        low = self.config.min_batch_size
        high = self.config.max_batch_size
        best = self.config.min_batch_size
        
        while low <= high:
            mid = low + (high - low) // 2
            total_memory = mid * memory_per_sample_mb
            
            if total_memory <= budget_mb:
                best = mid
                low = mid + 1
            else:
                high = mid - 1
        
        return max(self.config.min_batch_size, min(best, self.config.max_batch_size))
    
    def find_optimal_batch_size_binary_with_default(
        self,
        memory_per_sample_mb: float,
        default_batch_size: int,
    ) -> int:
        """Find optimal batch size with fallback default.
        
        Args:
            memory_per_sample_mb: Memory required per sample in MB
            default_batch_size: Batch size to return if auto-adjust is disabled
            
        Returns:
            The optimal batch size or default if auto-adjust is disabled
        """
        if not self.config.auto_adjust:
            return default_batch_size
        return self.find_optimal_batch_size_binary(memory_per_sample_mb)


# =============================================================================
# Device Selection Functions
# =============================================================================

def is_cuda_available() -> bool:
    """Check if CUDA is available."""
    return torch.cuda.is_available()


def is_mps_available() -> bool:
    """Check if MPS (Metal Performance Shaders) is available."""
    return hasattr(torch.backends, "mps") and torch.backends.mps.is_available()


def get_device(priority: DevicePriority = DevicePriority.CUDA_FIRST) -> torch.device:
    """Get the best available device with specified priority.
    
    Default priority: CUDA → MPS → CPU
    
    Args:
        priority: Device selection priority
        
    Returns:
        The best available torch.device
    """
    config = DeviceConfig(priority=priority)
    return get_device_with_config(config)


def get_device_with_config(config: DeviceConfig) -> torch.device:
    """Get device with custom configuration.
    
    Args:
        config: Device configuration
        
    Returns:
        The selected torch.device
    """
    if config.priority == DevicePriority.CPU_ONLY:
        logger.info("Using CPU (configured)")
        return torch.device("cpu")
    
    if config.priority == DevicePriority.CUDA_FIRST:
        return _get_cuda_first(config)
    else:
        return _get_mps_first(config)


def _get_cuda_first(config: DeviceConfig) -> torch.device:
    """CUDA → MPS → CPU priority chain."""
    if is_cuda_available():
        logger.info(f"Using CUDA GPU (device {config.cuda_device_id})")
        return torch.device(f"cuda:{config.cuda_device_id}")
    
    if is_mps_available():
        logger.info("Using MPS GPU - CUDA not available")
        return torch.device("mps")
    
    logger.info("Using CPU - no GPU available")
    return torch.device("cpu")


def _get_mps_first(config: DeviceConfig) -> torch.device:
    """MPS → CUDA → CPU priority chain."""
    if is_mps_available():
        logger.info("Using MPS GPU")
        return torch.device("mps")
    
    if is_cuda_available():
        logger.info(f"Using CUDA GPU (device {config.cuda_device_id}) - MPS not available")
        return torch.device(f"cuda:{config.cuda_device_id}")
    
    logger.info("Using CPU - no GPU available")
    return torch.device("cpu")


def get_device_info() -> dict[str, str]:
    """Get device availability information.
    
    Returns:
        Dictionary with device availability and selection info
    """
    cuda_available = is_cuda_available()
    mps_available = is_mps_available()
    
    if cuda_available:
        selected = "cuda"
    elif mps_available:
        selected = "mps"
    else:
        selected = "cpu"
    
    return {
        "cuda_available": str(cuda_available),
        "mps_available": str(mps_available),
        "selected_device": selected,
    }


def device_type_string(device: torch.device) -> str:
    """Get string representation of device type.
    
    Args:
        device: The torch device
        
    Returns:
        Device type as string ("cuda", "mps", or "cpu")
    """
    return device.type


# =============================================================================
# Convenience exports
# =============================================================================

__all__ = [
    # Enums
    "DevicePriority",
    # Config classes
    "DeviceConfig",
    "AutoBatchConfig",
    # Classes
    "AutoBatchSizer",
    # Functions
    "get_device",
    "get_device_with_config",
    "get_device_info",
    "device_type_string",
    "is_cuda_available",
    "is_mps_available",
]

File: rust-src/src/utils/device.rs
//! Unified Device Selection for DeepSeek Rust
//!
//! This module provides a centralized device selection mechanism with
//! CUDA → Metal → CPU priority chain.
//!
//! # Features
//! - Automatic device detection with consistent priority (CUDA first)
//! - Configurable device selection via environment variables
//! - Device information and availability checking
//!
//! # Example
//! ```rust,ignore
//! use deepseek_rust::utils::device::{DeviceSelector, DeviceConfig};
//!
//! // Get best available device (CUDA → Metal → CPU)
//! let device = DeviceSelector::get_device()?;
//!
//! // Or with custom configuration
//! let config = DeviceConfig::from_env();
//! let device = DeviceSelector::get_device_with_config(&config)?;
//! ```

use candle_core::{Device, Result};
use std::collections::HashMap;
use std::env;
use tracing::info;

/// Device selection priority
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DevicePriority {
    /// CUDA → Metal → CPU (recommended for training)
    CudaFirst,
    /// Metal → CUDA → CPU (for Apple Silicon optimization)
    MetalFirst,
    /// Always use CPU
    CpuOnly,
}

impl Default for DevicePriority {
    fn default() -> Self {
        Self::CudaFirst
    }
}

/// Configuration for device selection
#[derive(Debug, Clone)]
pub struct DeviceConfig {
    /// Device selection priority
    pub priority: DevicePriority,
    /// CUDA device ID (for multi-GPU systems)
    pub cuda_device_id: usize,
    /// Metal device ID
    pub metal_device_id: usize,
    /// Minimum batch size for auto-adjustment
    pub min_batch_size: usize,
    /// Maximum batch size for auto-adjustment
    pub max_batch_size: usize,
}

impl Default for DeviceConfig {
    fn default() -> Self {
        Self {
            priority: DevicePriority::CudaFirst,
            cuda_device_id: 0,
            metal_device_id: 0,
            min_batch_size: 1,
            max_batch_size: 256,
        }
    }
}

impl DeviceConfig {
    /// Create configuration from environment variables
    ///
    /// Environment variables:
    /// - `DEEPSEEK_DEVICE_PRIORITY`: "cuda", "metal", or "cpu"
    /// - `DEEPSEEK_CUDA_DEVICE`: CUDA device ID (default: 0)
    /// - `DEEPSEEK_METAL_DEVICE`: Metal device ID (default: 0)
    /// - `DEEPSEEK_MIN_BATCH_SIZE`: Minimum batch size (default: 1)
    /// - `DEEPSEEK_MAX_BATCH_SIZE`: Maximum batch size (default: 256)
    pub fn from_env() -> Self {
        let priority = match env::var("DEEPSEEK_DEVICE_PRIORITY")
            .unwrap_or_default()
            .to_lowercase()
            .as_str()
        {
            "metal" => DevicePriority::MetalFirst,
            "cpu" => DevicePriority::CpuOnly,
            _ => DevicePriority::CudaFirst, // Default
        };

        let cuda_device_id = env::var("DEEPSEEK_CUDA_DEVICE")
            .ok()
            .and_then(|s| s.parse().ok())
            .unwrap_or(0);

        let metal_device_id = env::var("DEEPSEEK_METAL_DEVICE")
            .ok()
            .and_then(|s| s.parse().ok())
            .unwrap_or(0);

        let min_batch_size = env::var("DEEPSEEK_MIN_BATCH_SIZE")
            .ok()
            .and_then(|s| s.parse().ok())
            .unwrap_or(1);

        let max_batch_size = env::var("DEEPSEEK_MAX_BATCH_SIZE")
            .ok()
            .and_then(|s| s.parse().ok())
            .unwrap_or(256);

        Self {
            priority,
            cuda_device_id,
            metal_device_id,
            min_batch_size,
            max_batch_size,
        }
    }

    /// Get the priority order as a list of device type names
    pub fn get_priority_order(&self) -> Vec<&'static str> {
        match self.priority {
            DevicePriority::CudaFirst => vec!["cuda", "metal", "cpu"],
            DevicePriority::MetalFirst => vec!["metal", "cuda", "cpu"],
            DevicePriority::CpuOnly => vec!["cpu"],
        }
    }
}

/// Centralized device selection with CUDA → Metal → CPU priority
pub struct DeviceSelector;

impl DeviceSelector {
    /// Get the best available device using default configuration (CUDA first)
    ///
    /// Priority: CUDA → Metal → CPU
    ///
    /// # Returns
    /// The best available device
    ///
    /// # Errors
    /// Returns error if device initialization fails (rare, usually GPU driver issues)
    pub fn get_device() -> Result<Device> {
        Self::get_device_with_config(&DeviceConfig::default())
    }

    /// Get device with custom configuration
    pub fn get_device_with_config(config: &DeviceConfig) -> Result<Device> {
        match config.priority {
            DevicePriority::CudaFirst => Self::get_cuda_first(config),
            DevicePriority::MetalFirst => Self::get_metal_first(config),
            DevicePriority::CpuOnly => {
                info!("Using CPU (configured)");
                Ok(Device::Cpu)
            }
        }
    }

    /// CUDA → Metal → CPU priority chain
    fn get_cuda_first(config: &DeviceConfig) -> Result<Device> {
        // Try CUDA first
        if candle_core::utils::cuda_is_available() {
            info!("Using CUDA GPU (device {})", config.cuda_device_id);
            return Device::new_cuda(config.cuda_device_id);
        }

        // Fallback to Metal
        if candle_core::utils::metal_is_available() {
            info!("Using Metal GPU (device {}) - CUDA not available", config.metal_device_id);
            return Device::new_metal(config.metal_device_id);
        }

        // Final fallback to CPU
        info!("Using CPU - no GPU available");
        Ok(Device::Cpu)
    }

    /// Metal → CUDA → CPU priority chain
    fn get_metal_first(config: &DeviceConfig) -> Result<Device> {
        // Try Metal first
        if candle_core::utils::metal_is_available() {
            info!("Using Metal GPU (device {})", config.metal_device_id);
            return Device::new_metal(config.metal_device_id);
        }

        // Fallback to CUDA
        if candle_core::utils::cuda_is_available() {
            info!("Using CUDA GPU (device {}) - Metal not available", config.cuda_device_id);
            return Device::new_cuda(config.cuda_device_id);
        }

        // Final fallback to CPU
        info!("Using CPU - no GPU available");
        Ok(Device::Cpu)
    }

    /// Check if CUDA is available
    pub fn is_cuda_available() -> bool {
        candle_core::utils::cuda_is_available()
    }

    /// Check if Metal is available
    pub fn is_metal_available() -> bool {
        candle_core::utils::metal_is_available()
    }

    /// Get device information as a HashMap
    pub fn get_device_info() -> HashMap<String, String> {
        let mut info = HashMap::new();
        
        info.insert(
            "cuda_available".to_string(),
            Self::is_cuda_available().to_string(),
        );
        info.insert(
            "metal_available".to_string(),
            Self::is_metal_available().to_string(),
        );

        // Determine selected device
        let selected = if Self::is_cuda_available() {
            "cuda"
        } else if Self::is_metal_available() {
            "metal"
        } else {
            "cpu"
        };
        info.insert("selected_device".to_string(), selected.to_string());

        info
    }

    /// Get string representation of device type
    pub fn device_type_string(device: &Device) -> &'static str {
        match device {
            Device::Cpu => "cpu",
            Device::Cuda(_) => "cuda",
            Device::Metal(_) => "metal",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_config() {
        let config = DeviceConfig::default();
        assert_eq!(config.priority, DevicePriority::CudaFirst);
        assert_eq!(config.cuda_device_id, 0);
        assert_eq!(config.metal_device_id, 0);
        assert_eq!(config.min_batch_size, 1);
        assert_eq!(config.max_batch_size, 256);
    }

    #[test]
    fn test_priority_order_cuda_first() {
        let config = DeviceConfig {
            priority: DevicePriority::CudaFirst,
            ..Default::default()
        };
        assert_eq!(config.get_priority_order(), vec!["cuda", "metal", "cpu"]);
    }

    #[test]
    fn test_priority_order_metal_first() {
        let config = DeviceConfig {
            priority: DevicePriority::MetalFirst,
            ..Default::default()
        };
        assert_eq!(config.get_priority_order(), vec!["metal", "cuda", "cpu"]);
    }

    #[test]
    fn test_priority_order_cpu_only() {
        let config = DeviceConfig {
            priority: DevicePriority::CpuOnly,
            ..Default::default()
        };
        assert_eq!(config.get_priority_order(), vec!["cpu"]);
    }

    #[test]
    fn test_get_device_returns_valid() {
        let device = DeviceSelector::get_device().expect("Should get device");
        // Verify device is usable
        let _tensor = candle_core::Tensor::zeros((2, 2), candle_core::DType::F32, &device)
            .expect("Should create tensor");
    }

    #[test]
    fn test_device_info_contains_required_keys() {
        let info = DeviceSelector::get_device_info();
        assert!(info.contains_key("cuda_available"));
        assert!(info.contains_key("metal_available"));
        assert!(info.contains_key("selected_device"));
    }

    #[test]
    fn test_device_type_string() {
        assert_eq!(DeviceSelector::device_type_string(&Device::Cpu), "cpu");
    }
}
